modin/experimental/core/execution/native/implementations/hdk_on_native/io/io.py

```diff
@@ -128,22 +129,54 @@ def read_csv(cls, **kwargs):  # noqa: PR01
             if not use_modin_impl:
                 raise ArrowEngineException(error_message)
 
-            dtype = kwargs["dtype"]
-            if isinstance(dtype, dict):
-                column_types = {c: cls._dtype_to_arrow(t) for c, t in dtype.items()}
-            else:
-                column_types = cls._dtype_to_arrow(dtype)
-
-            parse_dates = kwargs["parse_dates"]
-            if (type(parse_dates) is list) and type(column_types) is dict:
-                for c in parse_dates:
-                    column_types[c] = pa.timestamp("s")
-
-            names = kwargs["names"]
+            if (names := kwargs["names"]) is lib.no_default:
+                names = None
             skiprows = kwargs["skiprows"]
-            if names not in (lib.no_default, None) and kwargs["header"] == 0:
+            if names and kwargs["header"] == 0:
                 skiprows = skiprows + 1 if skiprows is not None else 1
 
+            column_names = None
+
+            def get_col_names():
+                nonlocal column_names
+                if column_names is None:
+                    # Using pandas to read the column names
+                    column_names = pandas.read_csv(
+                        kwargs["filepath_or_buffer"], nrows=0, engine="c"
+                    ).columns.tolist()
+                return column_names
+
+            if dtype := kwargs["dtype"]:
+                if isinstance(dtype, dict):
+                    column_types = {c: cls._dtype_to_arrow(t) for c, t in dtype.items()}
+                else:
+                    dtype = cls._dtype_to_arrow(dtype)
+                    column_types = {name: dtype for name in get_col_names()}
+            else:
+                column_types = {}
+
+            if parse_dates := kwargs["parse_dates"]:
+                if isinstance(parse_dates, list) and isinstance(
+                    parse_dates[0], (str, int)
+                ):
+                    # Pandas uses datetime64[ns] dtype for dates.
+                    timestamp_dt = pa.timestamp("ns")
+                    if names and isinstance(parse_dates[0], str):
+                        # If both `names` and column names in `parse_dates` are
+                        # specified, replace the column names with indices.
+                        parse_dates = [names.index(name) for name in parse_dates]
```

Comment:
the `column_types` dictionary has to map column names to their types, not column indices, right?

Reply:
Right and the indices are converted to the real column names below. The `names` parameter is used to override the column names. If a new name is specified in `parse_dates`, we can't just copy this name to `column_types`, because the parser will fail. We have to replace this name with the real column name. That's why, we put indices to `parse_dates` and then replace with real names.

Reply:
thanks, got it!